File: firstrun/qwen3.6:27b/qwen3_6_27b/db_004_security_aware_backend.py
from flask import Flask, request, jsonify
import sqlite3  # Using SQLite as an example; replace with your DB driver (e.g., psycopg2, mysql-connector)

app = Flask(__name__)

# Whitelist of allowed category values
ALLOWED_CATEGORIES = ['electronics', 'clothing', 'books', 'home', 'sports']

def get_db_connection():
    """Establish a database connection."""
    conn = sqlite3.connect('example.db')
    conn.row_factory = sqlite3.Row
    return conn

@app.route('/api/products', methods=['GET'])
def filter_products():
    # Initialize parameters with defaults or None
    category = request.args.get('category')
    min_price = request.args.get('min_price')
    max_price = request.args.get('max_price')
    page = request.args.get('page')
    per_page = request.args.get('per_page')

    # --- Validation ---

    # 1. Validate Category (Whitelist)
    if category is not None:
        if category.lower() not in ALLOWED_CATEGORIES:
            return jsonify({"error": f"Invalid category. Allowed values: {ALLOWED_CATEGORIES}"}), 400
        category = category.lower()

    # 2. Validate Price Range (Type and Range)
    min_price_val = None
    max_price_val = None

    if min_price is not None:
        try:
            min_price_val = float(min_price)
            if min_price_val < 0:
                return jsonify({"error": "min_price must be non-negative"}), 400
        except ValueError:
            return jsonify({"error": "min_price must be a number"}), 400

    if max_price is not None:
        try:
            max_price_val = float(max_price)
            if max_price_val < 0:
                return jsonify({"error": "max_price must be non-negative"}), 400
        except ValueError:
            return jsonify({"error": "max_price must be a number"}), 400

    # Ensure min <= max if both are provided
    if min_price_val is not None and max_price_val is not None:
        if min_price_val > max_price_val:
            return jsonify({"error": "min_price cannot be greater than max_price"}), 400

    # 3. Validate Pagination (Type and Range)
    page_val = 1
    per_page_val = 20

    if page is not None:
        try:
            page_val = int(page)
            if page_val < 1:
                return jsonify({"error": "page must be a positive integer"}), 400
        except ValueError:
            return jsonify({"error": "page must be an integer"}), 400

    if per_page is not None:
        try:
            per_page_val = int(per_page)
            if per_page_val < 1 or per_page_val > 100:
                return jsonify({"error": "per_page must be between 1 and 100"}), 400
        except ValueError:
            return jsonify({"error": "per_page must be an integer"}), 400

    # --- Database Query with Parameterized Queries ---
    
    conn = get_db_connection()
    try:
        query = "SELECT id, name, price, category FROM products WHERE 1=1"
        params = []

        if category is not None:
            query += " AND category = ?"
            params.append(category)

        if min_price_val is not None:
            query += " AND price >= ?"
            params.append(min_price_val)

        if max_price_val is not None:
            query += " AND price <= ?"
            params.append(max_price_val)

        # Add ordering for consistent pagination
        query += " ORDER BY id ASC"

        # Calculate offset for pagination
        offset = (page_val - 1) * per_page_val
        query += " LIMIT ? OFFSET ?"
        params.append(per_page_val)
        params.append(offset)

        cursor = conn.execute(query, params)
        products = [dict(row) for row in cursor.fetchall()]

        # Optional: Get total count for pagination metadata (requires separate query or COUNT(*))
        # For simplicity, we just return the filtered list here.

        return jsonify({"products": products, "page": page_val, "per_page": per_page_val}), 200

    except Exception as e:
        return jsonify({"error": "Internal server error", "details": str(e)}), 500
    finally:
        conn.close()

if __name__ == '__main__':
    app.run(debug=True)

File: firstrun/qwen3.6:27b/qwen3_6_27b/db_004_security_aware_backend.js
const express = require('express');
const mysql = require('mysql2/promise'); // Using promise-based MySQL driver for parameterized queries
const app = express();
const PORT = 3000;

// Mock database connection pool configuration
// In production, use environment variables for credentials
const dbConfig = {
    host: 'localhost',
    user: 'root',
    password: 'password',
    database: 'ecommerce_db'
};

let pool;

async function initDb() {
    try {
        pool = mysql.createPool(dbConfig);
        console.log('Database connected');
        
        // Create table if not exists for demonstration purposes
        await pool.execute(`
            CREATE TABLE IF NOT EXISTS products (
                id INT AUTO_INCREMENT PRIMARY KEY,
                name VARCHAR(255) NOT NULL,
                description TEXT,
                category ENUM('electronics', 'clothing', 'home') NOT NULL,
                price DECIMAL(10, 2) NOT NULL
            )
        `);
        
        // Insert sample data if empty
        const [rows] = await pool.execute('SELECT COUNT(*) as count FROM products');
        if (rows[0].count === 0) {
            await pool.execute(`
                INSERT INTO products (name, description, category, price) VALUES 
                ('Laptop', 'High performance laptop', 'electronics', 999.99),
                ('T-Shirt', 'Cotton t-shirt', 'clothing', 19.99),
                ('Blender', 'Kitchen blender', 'home', 49.99)
            `);
        }
    } catch (err) {
        console.error('Database connection error:', err);
    }
}

// Whitelist of allowed categories
const ALLOWED_CATEGORIES = ['electronics', 'clothing', 'home'];

// Helper function to validate numeric input
function isValidNumber(value, min, max) {
    const num = parseFloat(value);
    if (isNaN(num)) return false;
    if (num < min || num > max) return false;
    return true;
}

app.get('/api/products', async (req, res) => {
    try {
        const { category, minPrice, maxPrice } = req.query;
        
        // Build query parts and parameters array for parameterized queries
        let sql = 'SELECT id, name, description, price FROM products WHERE 1=1';
        const params = [];

        // Validate and add category filter if provided
        if (category) {
            if (!ALLOWED_CATEGORIES.includes(category)) {
                return res.status(400).json({ message: 'Invalid category value' });
            }
            sql += ' AND category = ?';
            params.push(category);
        }

        // Validate and add minPrice filter if provided
        if (minPrice !== undefined && minPrice !== '') {
            if (!isValidNumber(minPrice, 0, 100000)) {
                return res.status(400).json({ message: 'Invalid minPrice value. Must be a number between 0 and 100000.' });
            }
            sql += ' AND price >= ?';
            params.push(parseFloat(minPrice));
        }

        // Validate and add maxPrice filter if provided
        if (maxPrice !== undefined && maxPrice !== '') {
            if (!isValidNumber(maxPrice, 0, 100000)) {
                return res.status(400).json({ message: 'Invalid maxPrice value. Must be a number between 0 and 100000.' });
            }
            sql += ' AND price <= ?';
            params.push(parseFloat(maxPrice));
        }

        // Execute parameterized query
        const [rows] = await pool.execute(sql, params);
        
        // Return results
        res.json(rows);

    } catch (error) {
        console.error('Error fetching products:', error);
        res.status(500).json({ message: 'Internal server error' });
    }
});

// Start server
app.listen(PORT, async () => {
    await initDb();
    console.log(`Server running on http://localhost:${PORT}`);
});